### PR TITLE
Upstream Triton Internal Review

### PR DESCRIPTION
**Pull request by micmelesse**
DON'T Merge. This is meant for AMD's Triton Kernels team to review.

Hi, this is a pr to add a Triton backend to Flash Attention on ROCm. We hope that this pr will be the first in a series of prs to that end. Triton has had support for ROCm for a while now and a Flash Attention Triton backend will allows us to support Flash Attention on both our MI and Navi Machines.

In this pr, we enable major parts of `fwd`, `varlen_fwd` and `fwd_kvcache`. However there are some features missing such as Dropout, Sliding window, Rotary Embedding and Pagged Attention. There are also a few miscellaneous bugs. These will all be addressed in subsequent prs. The next pr we plan to file will be support for `bwd` and `varlen_vwd`, if we should reprioritize, please let us know.

We have tested this pr [here](https://github.com/ROCm/flash-attention/actions/runs/10688682354/job/29629059809) on an MI200 machine. When the testing the Triton Backend for ROCm, we skip testing the backward pass, configs with unsupported features and a portion of  headsizes (d) randomly. The later is to keep the test times reasonable. The latest results, we have are `=== 64 failed, 30387 passed, 478321 skipped, 1 warning in 3110.86s (0:51:50) ===`.  There is clearly more work to be done but we hope that this will make a good start. 

Please let us know what we can do on our end to help with this process.
Finally this pr includes work from multiple people besides myself, especially thanks to @vgokhale, @scxiao and @jlgreathouse.

### CODE REVIEW

File: tests/test_flash_attn.py [hunk 1]
```diff
@@ -17,6 +19,31 @@
...
+def skip_config(*args,  skip_pct = 0.95):
+    return random.random() >= (1.0 - skip_pct)
+
+def skip_not_power_of_2_config(*args):
+    if not all(is_power_of_2(arg) for arg in args):
+        return True
+    else:
+        return False
+
+def is_amd():
```
Comment (scxiao) [inline tests/test_flash_attn.py:42]: Can we rename this to `is_hip()`, so be consistent with other places?

Comment (micmelesse) [reply tests/test_flash_attn.py]: done

File: flash_attn/flash_attn_triton_kernel_decode_amd.py [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+    V_block_ptr = tl.make_block_ptr(
+        base=v_base + stride_vd * QUANTIZED * N_QUANT_GROUPS,
+        shape=(hi, ACTUAL_BLOCK_DMODEL),
+        strides=(stride_vn, stride_vd),
+        offsets=(lo, 0),
+        block_shape=(BLOCK_N, BLOCK_DMODEL),
+        order=(1, 0),
+    )
+
+    if QUANTIZED:
```
Comment (scxiao) [inline flash_attn/flash_attn_triton_kernel_decode_amd.py:253]: I remember you mentioned you would remove the quantized part, right?

Comment (micmelesse) [reply flash_attn/flash_attn_triton_kernel_decode_amd.py:253]: I will remove the code. I was just asserting to ensure it was not triggered

**Pull request closed without merging**